### PR TITLE
Add EIP: On-Chain EIP-712 Visualization

### PR DESCRIPTION
**Pull request by a6-dou**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-6865.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
+eip: 6865
+title: on-chain EIP-712 Visualization
+description: on-chain visualization of EIP-712 payloads into structured data that highlights the potential consequences for users' assets
+author: Abderrahmen Hanafi (@a6-dou)
+discussions-to: https://ethereum-magicians.org/
```
Comment (lightclient) [inline EIPS/eip-6865.md:6]: please create an actual post on `ethereum-magicians` and use as your discussions to link

Comment (a6-dou) [reply EIPS/eip-6865.md]: Thanks for the review, [Pushed the update](https://github.com/ethereum/EIPs/pull/6865/commits/46fb7a3884be877e52601359e4a56add54372c48) with discussion link

**Review by cynic-1 — COMMENTED** [review]
LGTM

File: EIPS/eip-6865.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+---
+eip: 6865
+title: on-chain EIP-712 Visualization
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:3]:
```suggestion
title: On-Chain EIP-712 Visualization
```

File: EIPS/eip-6865.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+with visualization we can achieve something similar to image below where more insightful details are revealed to user thanks to the structured data returned from the EIP
+
+![](../assets/eip-6865/vision-EIP-712-signature-wallet-interface.png)
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Each new protocol that uses EIP-712 MUST include the `visualizeEIP712Message` function in the  `verifyingContract` implementation
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:44]: This proposal can't introduce restrictions on _all_ contracts implementing EIP-712. You could do something like:

```suggestion
Contracts implementing this proposal MUST include the `visualizeEIP712Message` function in the  `verifyingContract` implementation
```

File: EIPS/eip-6865.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+### Params
+
+`encodedMessage` is bytes that represents the encoded EIP-712  message with `abi.encode` and it can be decoded using `abi.decode`
+
+`domainHash` is the bytes32 hash of the EIP-712 domain separator as defined in the EIP-712 proposal
+
+### Outputs
+
+The function MUST returns Result ; a struct that contains information's about user’s assets impact and the liveness of such a message if get signed.
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:68]:
```suggestion
The function MUST return `Result`, a struct that contains information's about user’s assets impact and the liveness of such a message if it gets signed.
```

File: EIPS/eip-6865.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+- `amounts` is an Array of `uint256` where items MUST define the amount per time curve, with time defined within liveness boundaries
+    - the first amount in `amounts` Array (amounts[0]) MUST be the amount of the asset at `liveness.from` timestamp
+    - the last amount in `amounts` Array (amounts[amounts.length-1]) MUST be the amount of the asset at `liveness.to` timestamp
+    - in most of the cases, `amounts` will be an Array with a single item which is MUST be the minimum amount of the asset.
+
+#### assetsIn
+
+`assetsIn` are the minimum assets which the user MUST get if the message is signed and fulfilled
+
+#### assetsOut
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:113]: For all of the headers including code, please surround the code in backticks (\`)

Comment (a6-dou) [reply EIPS/eip-6865.md:113]: noted with thanks!

File: EIPS/eip-6865.md (hunk 2 above)
Comment (a6-dou) [reply EIPS/eip-6865.md]: Fixed in commit [573ff5b1ebf](https://github.com/ethereum/EIPs/pull/6865/commits/573ff5b1ebfb3329d5e9b404d5ef8b47c78a58f3)

File: EIPS/eip-6865.md (hunk 5 above)
Comment (a6-dou) [reply EIPS/eip-6865.md]: Fixed in commit [573ff5b1ebf](https://github.com/ethereum/EIPs/pull/6865/commits/573ff5b1ebfb3329d5e9b404d5ef8b47c78a58f3)

File: EIPS/eip-6865.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
+---
+eip: 6865
+title: on-chain EIP-712 Visualization
+description: on-chain visualization of EIP-712 payloads into structured data that highlights the potential consequences for users' assets
```
[the remark this answers is not in the document]
Comment (a6-dou) [reply EIPS/eip-6865.md]: fixed in commit [bc7c1d16dc](https://github.com/ethereum/EIPs/pull/6865/commits/bc7c1d16dcf5ac1ca6dbb683430b3e82f53af0e9)

File: EIPS/eip-6865.md (hunk 3 above)
Comment (a6-dou) [reply EIPS/eip-6865.md]: fixed in commit [bc7c1d16dc](https://github.com/ethereum/EIPs/pull/6865/commits/bc7c1d16dcf5ac1ca6dbb683430b3e82f53af0e9)

File: EIPS/eip-6865.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+To overcome these challenges, we propose a standardized, on-chain solution that can accommodate the diverse and ever-changing web3.0 ecosystem. This approach would enhance scalability, reliability, and maintainability by shifting the responsibility of visualizing EIP-712 payloads from the wallets to the protocols themselves. Consequently, wallets can use a consistent and effective approach to EIP-712 message visualization.
+
+The adoption of a universal solution will not only streamline the efforts and reduce the maintenance burden for wallet providers, but it will also allow for faster and more extensive coverage across the ecosystem. This will ultimately result in users gaining a clearer understanding of the transactions they're signing, leading to increased security and an improved overall user experience within the crypto space.
+
+currently, most of the wallets display something similar to image below
+
+![](../assets/eip-6865/current-EIP-712-signature-wallet-interface.png)
+
+with visualization we can achieve something similar to image below where more insightful details are revealed to user thanks to the structured data returned from the EIP
```
[the remark this answers is not in the document]
Comment (a6-dou) [reply EIPS/eip-6865.md]: fixed in commit [92b1ab4fb](https://github.com/ethereum/EIPs/pull/6865/commits/92b1ab4fb061db98e5a2f166034dba633e6ee297)

File: EIPS/eip-6865.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+- Maintainability: Offering support for protocols with an off-chain solution is insufficient in a rapidly evolving ecosystem. Protocols frequently upgrade their implementations by extending features or fixing bugs, further complicating the maintenance process.
+
+To overcome these challenges, we propose a standardized, on-chain solution that can accommodate the diverse and ever-changing web3.0 ecosystem. This approach would enhance scalability, reliability, and maintainability by shifting the responsibility of visualizing EIP-712 payloads from the wallets to the protocols themselves. Consequently, wallets can use a consistent and effective approach to EIP-712 message visualization.
+
+The adoption of a universal solution will not only streamline the efforts and reduce the maintenance burden for wallet providers, but it will also allow for faster and more extensive coverage across the ecosystem. This will ultimately result in users gaining a clearer understanding of the transactions they're signing, leading to increased security and an improved overall user experience within the crypto space.
+
+currently, most of the wallets display something similar to image below
```
[the remark this answers is not in the document]
Comment (a6-dou) [reply EIPS/eip-6865.md]: fixed in commit [92b1ab4fb](https://github.com/ethereum/EIPs/pull/6865/commits/92b1ab4fb061db98e5a2f166034dba633e6ee297)

File: EIPS/eip-6865.md (hunk 4 above)
Comment (a6-dou) [reply EIPS/eip-6865.md]: fixed in commit [92b1ab4fb](https://github.com/ethereum/EIPs/pull/6865/commits/92b1ab4fb061db98e5a2f166034dba633e6ee297)

File: EIPS/eip-6865.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+struct Result {
+  UserAssetMovement[] assetsIn;
+  UserAssetMovement[] assetsOut;
+  Liveness liveness;
+}
+```
+
+#### `Liveness`
+
+`liveness` is a struct that defines the timestamps which the message is valid where:
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:91]:
```suggestion
`Liveness` is a struct that defines the timestamps which the message is valid where:
```

File: EIPS/eip-6865.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+struct Liveness {
+  uint256 from;
+	uint256 to;
+}
+
+struct UserAssetMovement {
+	address assetTokenAddress;
+  uint256 id;
+  uint256[] amounts;
+}
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:80]:
```suggestion
struct UserAssetMovement {
  address assetTokenAddress;
  uint256 id;
  uint256[] amounts;
}
```

File: EIPS/eip-6865.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+`domainHash` is the bytes32 hash of the EIP-712 domain separator as defined in the EIP-712 proposal
+
+### Outputs
+
+The function MUST return `Result`, a struct that contains information's about user’s assets impact and the liveness of such a message if it gets signed.
+
+```solidity
+struct Liveness {
+  uint256 from;
+	uint256 to;
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:73]:
```suggestion
  uint256 from;
  uint256 to;
```

File: assets/eip-6865/contracts/SeaPortEIP712Visualizer.sol [hunk 12]
```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MIT
```
Comment (SamWilsn) [inline assets/eip-6865/contracts/SeaPortEIP712Visualizer.sol:1]: Is SeaPort licensed as MIT?

File: EIPS/eip-6865.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+No backward compatibility issues found.
+
+## Reference Implementation
+
+openSea Seaport NFT marketplace implementation example is available [here](../assets/eip-6865/contracts/SeaPortEIP712Visualizer.sol)
+
+## Security Considerations
+
+`visualizeEIP712Message` function should be reliable and accurately represent the potential impact of the EIP-712 message on users' assets. Wallet providers and users must trust the protocol's implementation of this function to provide accurate and up-to-date information.
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:147]: Normally I wouldn't hold up a proposal because of a security consideration, but I think this needs to be addressed even before going into draft (because proposals in draft sometimes stay in draft even after implementations are released...)

What happens if I'm transacting with a malicious website `obviousscam.invalid`, and `obviousscam.invalid` pops up a sign typed data request. What in this proposal stops `obviousscam.invalid`'s contract from showing a completely different visualization from what the real contract will do?

If it's the `verifyingContract` in the domain separator, you should mention that signing requests without that field do not get visualized.

File: EIPS/eip-6865.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+With visualization we can achieve something similar to image below where more insightful details are revealed to user thanks to the structured data returned from the EIP
+
+![](../assets/eip-6865/vision-EIP-712-signature-wallet-interface.png)
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Contracts implementing this proposal MUST include the `visualizeEIP712Message` function in the  `verifyingContract` implementation
+
```
Comment (SamWilsn) [inline EIPS/eip-6865.md:45]: Even though this is a technical comment, I'm putting it in this pull request instead of in the discussions-to thread because without an answer in the proposal, I don't feel this specification is complete enough:

How should a wallet link a signature request to a contract? Like if I call `eth_signTypedData`, how does the wallet know which contract to use for `visualizeEIP712Message`? The `verifyingContract` in the domain separator? What if the domain separator doesn't include that field?

File: EIPS/eip-6865.md (hunk 13 above)
Comment (a6-dou) [reply EIPS/eip-6865.md:147]: also, users usually and wallets SHOULD always highlight to user the nature and reputation of the smart contract they are interacting with, let's take an example to make it clear:
user head up to `obviousscam.invalid` where this scammy website is trying to steal the funds, and i can imagine two possible scenarios using EIP-712:
## 1
`obviousscam.invalid` ask user to sign an EIP-712 where the `verifyingContract` is `0xbad` which is the scam website smart contract itself, at this stage user didn't give any approval to `0xbad` or any address associated with this entity, because if s/he did, there's no need even to reach this point because the scammy entity would have already rekt the user.
means even if the `0xbad` lies about it's behaviour to scam user, there's no real impact at that time because the EIP-712 message is intended to that entity itself only, and cant be used elsewhere.
indeed, to even cut it at that level, wallets usually(and SHOULD always) show the verified sign when it comes to smart contract addresses, so a known protocol smart contract like openSea, looksrare, blur, curve..etc are displayed as verified/trusted smart contract (code is low), but less-knows to unknown protocols/smart contracts are always displayed as raw addresses and users usually (should be) warned that this smart contract is unknown and s/he must double check what s/he is doing (this applies to all crypto space, even RPC simulation can be misleading). 

##2 
`obviousscam.invalid` ask user to sign an EIP-712 where the `verifyingContract` is some other protocol, for example opensea marketplace (which is the case of many phishing attacks , you can check [this $1.07 million attack](https://twitter.com/Serpent/status/1604074440941506560?lang=en)
in this case wallets uses `verifyingContract` which points to the real opensea protocol address, they show that this EIP-712 message is intended for opensea (which is already suspicious to user), wallet call [this proposal](https://github.com/ethereum/EIPs/pull/6865) smart contact implementation and gets the `Result` that consist of `assetsIn` and `assetsOut` which can be easily displayed/analysed on the front-side 

TLDR; `verifyingContract` should be displayed based on the address reputation which is something already in place for every smart contract (on-chain transactions), and the EIP-712 message should be treated based on the reputation of its `verifyingContract`, any EIP-712 message that doesn't have a `verifyingContract` is ignored and honestly i don't see a case of an EIP-712 message that needs to be executed on-chain without the `verifyingContract`

Comment (a6-dou) [reply EIPS/eip-6865.md:147]: will update the PR accordingly and request you review again, Thanks!

Comment (a6-dou) [reply EIPS/eip-6865.md:149]: add more details [here](https://github.com/ethereum/EIPs/pull/6865/commits/d712f865c9cd9ff18144cd1c72d94c8068799043)

File: EIPS/eip-6865.md (hunk 14 above)
Comment (a6-dou) [reply EIPS/eip-6865.md:47]: add more details [here](https://github.com/ethereum/EIPs/pull/6865/commits/928ff07cdbc8bcf82a1491fdcfacef6be1471e95)

**Review by SamWilsn — APPROVED** [review]
